rfcs/QueryLevelNullability.md

```diff
@@ -0,0 +1,178 @@
+This is in contrast to the two modern languages used on mobile clients, Swift and Kotlin,
+which are both non-null by default. In Swift and Kotlin, unless developers otherwise
+specify it, properties cannot be `null`.
+
+This mismatch creates some dissonance for developers who are currently forced into dealing the
+problems that commonly surround nullablility in codebases that otherwise do not need to deal with
+those problems. This makes large numbers of nulls tedious and time-consuming to handle.
+
+It is often unclear how to handle partial results.
+- What elements are essential to having an adequate user experience? How many elements can fail before you 
+  should just show an error message instead?
+- When any combination of elements can fail, it is very hard to anticipate all edge cases and how the app 
+  might look and work, including transitions to other screens.
+
+In Yelp's GraphQL schema, almost all object fields are nullable except for those with ID type. 
+This adheres to what seems to be the [official best practice](https://graphql.org/learn/best-practices/#nullability).
+
+This poses a problem for the mobile clients that use Apollo's codegen feature. The codegen provides 
+Swift/Kotlin types to represent queries used in the app. Nullable fields in the schema are represented
+as nullable properties on the resulting type, represented by `?` following the type name:
+```graphql
+query GetBusinessName($encid: String!) {
+  business(encid: $encid) {
+    name
+  }
+}
+```
+```swift
+struct GetBusinessNameQuery.Data.Business {
+  let name: String?
+}
+```
+In many cases, the client should error if the business `name` is null. If codegen were out of the picture,
+we would be able to throw an error at JSON response-parsing time if it's missing, or otherwise instantiate
+a hand-written business object with a non-nullable `name` property. From that point on, all feature code
+can happily assume that it has a non-null business name to work with.
+
+## 🧑‍💻 Proposed syntax
+
+It would make more sense if the client could express that `name` must be non-null _in the query itself_:
+```graphql
+query GetBusinessName($encid: String!) {
+  business(encid: $encid) {
+    name!  <-- this!
+  }
+}
+```
+Semantically the GraphQL `!` operator is nearly identical to its counterpart in Swift (also represented by `!`) which is
+referred to as the "force unwrap operator". In Swift, for example, you can cast a string to an integer with `Int("5")` 
+but the string being cast may not be a valid number, so that statement will return `null` rather than an integer if the
+string cannot be turned into an integer. If you want to ensure that the statement does not return `null` you can instead 
+write `Int("5")!`. If you do that, an exception will be thrown if the statement would return `null`.
+
+In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an
+error to the client rather than any data.
```

Comment:
What kind of error is best in this situation? Should that be at the discretion of the user rather than being baked into the spec?